### PR TITLE
Support even more multiple outlets 

### PR DESCRIPTION
**Pull request by mrsteakhouse**
Good day,

This is based on this PR: https://github.com/WebThingsIO/zigbee-adapter/pull/124
In regards to this issue https://github.com/WebThingsIO/zigbee-adapter/issues/236 i made some changes to additionally support multiple endpoints on a single device. Also many thanks to @dhylands for doing the hard work.
Tested with my local Gateway setup and the [Innr RC 110](https://www.innr.com/de/produkt/remote-control/).

When setting the remote to scenes, the on/off button and +/- buttons work, however i did not manage to get the numbered buttons to work.
Setting the slider to lights enables the numbered buttons to alter the codes for the top 3 buttons. Each now controls a different property with each on/off and level state.
Anyways i'm very happy with the result.

I'm not very fluent with JS so please point out any bad parts.

The classifier log messages:
```
020-11-02 20:31:57.276 INFO   : zigbee-adapter: ---- Zigbee classifier -----
2020-11-02 20:31:57.276 INFO   : zigbee-adapter:                    modelId = RC 110
2020-11-02 20:31:57.277 INFO   : zigbee-adapter:         seMeteringEndpoint = undefined
2020-11-02 20:31:57.277 INFO   : zigbee-adapter:       haElectricalEndpoint = undefined
2020-11-02 20:31:57.283 INFO   : zigbee-adapter:     genBinaryInputEndpoint = undefined
2020-11-02 20:31:57.283 INFO   : zigbee-adapter:       genLevelCtrlEndpoint = undefined
2020-11-02 20:31:57.283 INFO   : zigbee-adapter: genLevelCtrlOutputEndpoint = [
2020-11-02 20:31:57.283 INFO   : zigbee-adapter:   1, 3, 4, 5,
2020-11-02 20:31:57.284 INFO   : zigbee-adapter:   6, 7, 8
2020-11-02 20:31:57.284 INFO   : zigbee-adapter: ]
2020-11-02 20:31:57.284 INFO   : zigbee-adapter:          genOnOffEndpoints = []
2020-11-02 20:31:57.285 INFO   : zigbee-adapter:     genOnOffOutputEndpoint = [
2020-11-02 20:31:57.285 INFO   : zigbee-adapter:   1, 3, 4, 5,
2020-11-02 20:31:57.285 INFO   : zigbee-adapter:   6, 7, 8
2020-11-02 20:31:57.285 INFO   : zigbee-adapter: ]
2020-11-02 20:31:57.286 INFO   : zigbee-adapter:     hvacFanControlEndpoint = undefined
2020-11-02 20:31:57.286 INFO   : zigbee-adapter:     hvacThermostatEndpoint = undefined
2020-11-02 20:31:57.286 INFO   : zigbee-adapter:           doorLockEndpoint = undefined
2020-11-02 20:31:57.286 INFO   : zigbee-adapter:     lightingContolEndpoint = undefined
2020-11-02 20:31:57.287 INFO   : zigbee-adapter:          colorCapabilities = undefined
2020-11-02 20:31:57.287 INFO   : zigbee-adapter:                  colorMode = undefined
2020-11-02 20:31:57.287 INFO   : zigbee-adapter:          lightLinkEndpoint = 2
2020-11-02 20:31:57.287 INFO   : zigbee-adapter: msOccupancySensingEndpoint = undefined
2020-11-02 20:31:57.288 INFO   : zigbee-adapter:      msTemperatureEndpoint = undefined
2020-11-02 20:31:57.288 INFO   : zigbee-adapter:        genPowerCfgEndpoint = undefined
2020-11-02 20:31:57.288 INFO   : zigbee-adapter:   genDeviceTempCfgEndpoint = undefined
2020-11-02 20:31:57.288 INFO   : zigbee-adapter:                   zoneType = undefined
2020-11-02 20:31:57.289 INFO   : zigbee-adapter: Processing endpoint 0 = 1
2020-11-02 20:31:57.289 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on0 EP:1 CL:0006
2020-11-02 20:31:57.289 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.289 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.290 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level0 EP:1 CL:0008
2020-11-02 20:31:57.294 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.294 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.295 INFO   : zigbee-adapter: Processing endpoint 1 = 3
2020-11-02 20:31:57.295 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on1 EP:3 CL:0006
2020-11-02 20:31:57.295 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.295 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.296 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level1 EP:3 CL:0008
2020-11-02 20:31:57.296 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.297 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.298 INFO   : zigbee-adapter: Processing endpoint 2 = 4
2020-11-02 20:31:57.298 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on2 EP:4 CL:0006
2020-11-02 20:31:57.299 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.299 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.300 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level2 EP:4 CL:0008
2020-11-02 20:31:57.300 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.301 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.302 INFO   : zigbee-adapter: Processing endpoint 3 = 5
2020-11-02 20:31:57.303 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on3 EP:5 CL:0006
2020-11-02 20:31:57.304 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.305 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.305 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level3 EP:5 CL:0008
2020-11-02 20:31:57.306 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.306 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.307 INFO   : zigbee-adapter: Processing endpoint 4 = 6
2020-11-02 20:31:57.308 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on4 EP:6 CL:0006
2020-11-02 20:31:57.308 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.309 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.309 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level4 EP:6 CL:0008
2020-11-02 20:31:57.310 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.310 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.311 INFO   : zigbee-adapter: Processing endpoint 5 = 7
2020-11-02 20:31:57.312 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on5 EP:7 CL:0006
2020-11-02 20:31:57.313 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.313 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.314 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level5 EP:7 CL:0008
2020-11-02 20:31:57.314 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.316 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.317 INFO   : zigbee-adapter: Processing endpoint 6 = 8
2020-11-02 20:31:57.319 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 on6 EP:8 CL:0006
2020-11-02 20:31:57.320 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.320 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: false
2020-11-02 20:31:57.321 INFO   : zigbee-adapter: addProperty: 00158d00017ced47 level6 EP:8 CL:0008
2020-11-02 20:31:57.322 INFO   : zigbee-adapter: addProperty:    fireAndForget: true bindNeeded: false configReportNeeded: false initialReadNeeded: false
2020-11-02 20:31:57.322 INFO   : zigbee-adapter: addProperty:   bindNeeded: true value: 0
2020-11-02 20:31:57.323 INFO   : zigbee-adapter: classify: Initialized as type: undefined name: zb-00158d00017ced47-button defaultName: zb-00158d00017ced47-undefined
```

### CODE REVIEW

File: zb-classifier.js [hunk 1]
```diff
@@ -1778,13 +1802,14 @@ class ZigbeeClassifier {
...
-      this.initOnOffSwitch(node, genOnOffEndpoint);
-    } else if (genLevelCtrlOutputEndpoint) {
-      this.initMultiLevelButton(node, genLevelCtrlOutputEndpoint,
-                                genOnOffOutputEndpoint);
-    } else if (genOnOffOutputEndpoint) {
-      this.initOnOffButton(node, genOnOffOutputEndpoint);
+    } else if (genOnOffEndpoints.length > 0) {
+      this.initOnOffSwitches(node, genOnOffEndpoints);
+    } else if (genLevelCtrlOutputEndpoints.length > 0 &&
+      genOnOffOutputEndpoints.length > 0) {
```
Comment (mrstegeman) [inline zb-classifier.js:1808]: nit: spacing should align with the previous line

File: zb-classifier.js [hunk 2]
```diff
@@ -469,15 +472,30 @@ class ZigbeeClassifier {
...
 
-  addButtonOnProperty(node, genOnOffOutputEndpoint) {
+  addButtonOnProperty(node, genOnOffOutputEndpoint, suffix) {
+    if (typeof suffix === 'undefined') {
+      suffix = '';
+    }
     const property = this.addProperty(
       node,                           // device
-      'on',                           // name
+      `on${suffix}`,                           // name
```
Comment (mrstegeman) [inline zb-classifier.js:495]: nit: comment spacing should align with previous line

File: zb-classifier.js [hunk 3]
```diff
@@ -499,13 +517,16 @@ class ZigbeeClassifier {
...
 
-  addButtonLevelProperty(node, genLevelCtrlOutputEndpoint) {
+  addButtonLevelProperty(node, genLevelCtrlOutputEndpoint, suffix) {
+    if (typeof suffix === 'undefined') {
+      suffix = '';
+    }
     const property = this.addProperty(
       node,                           // device
-      'level',                        // name
+      `level${suffix}`,                        // name
```
Comment (mrstegeman) [inline zb-classifier.js:526]: nit: comment spacing should align with previous line

File: zb-classifier.js [hunk 4]
```diff
@@ -531,14 +552,17 @@ class ZigbeeClassifier {
...
 
-  addButtonMotionProperty(node, genOnOffOutputEndpoint) {
+  addButtonMotionProperty(node, genOnOffOutputEndpoint, suffix) {
+    if (typeof suffix === 'undefined') {
+      suffix = '';
+    }
     const property = this.addProperty(
       node,                           // device
-      'motion',                       // name
+      `motion${suffix}`,                       // name
```
Comment (mrstegeman) [inline zb-classifier.js:561]: nit: comment spacing should align with previous line

File: zb-classifier.js [hunk 5]
```diff
@@ -1738,9 +1762,9 @@ class ZigbeeClassifier {
       console.log('      haElectricalEndpoint =', haElectricalEndpoint);
       console.log('    genBinaryInputEndpoint =', genBinaryInputEndpoint);
       console.log('      genLevelCtrlEndpoint =', genLevelCtrlEndpoint);
-      console.log('genLevelCtrlOutputEndpoint =', genLevelCtrlOutputEndpoint);
-      console.log('          genOnOffEndpoint =', genOnOffEndpoint);
-      console.log('    genOnOffOutputEndpoint =', genOnOffOutputEndpoint);
+      console.log('genLevelCtrlOutputEndpoint =', genLevelCtrlOutputEndpoints);
```
Comment (mrstegeman) [inline zb-classifier.js:1765]: nit: you pluralized the label for one of these, but not the other two

File: zb-classifier.js [hunk 6]
```diff
@@ -1987,161 +2017,177 @@ class ZigbeeClassifier {
...
+              '@type': 'LongPressedEvent',
+              description: `${label[i]} button pressed and held`,
+            },
+            [`${i + 2}-released`]: {
+              '@type': 'ReleasedEvent',
+              description: `${label[i]} button released (after being held)`,
+            },
+          });
+        }
+        continue;
```
Comment (mrstegeman) [inline zb-classifier.js:2135]: This should remain `return` to prevent broken Tradfri remote support.

File: zb-classifier.js [hunk 7]
```diff
@@ -1987,161 +2017,177 @@ class ZigbeeClassifier {
...
+          },
+        });
+      }
     }
   }
 
-  initMultiLevelButton(node, genLevelCtrlOutputEndpoint,
-                       genOnOffOutputEndpoint) {
-    if (node.modelId.includes('motion') && genOnOffOutputEndpoint) {
+  initMultiLevelButton(node, genLevelCtrlOutputEndpoints,
```
Comment (mrstegeman) [inline zb-classifier.js:2060]: Should this be `initMultiLevelButtons` now?

File: zb-classifier.js [hunk 8]
```diff
@@ -1987,161 +2017,177 @@ class ZigbeeClassifier {
     this.addOnProperty(node, genLevelCtrlEndpoint);
   }
 
-  initOnOffButton(node, genOnOffOutputEndpoint) {
-    if (node.modelId.includes('motion')) {
-      // The IKEA Motion sensor has a modelId of 'TRADFRI motion sensor'
-      node.name = `${node.id}-motion`;
-      node['@type'] = ['MotionSensor'];
+  initOnOffButton(node, genOnOffOutputEndpoints) {
```
Comment (mrstegeman) [inline zb-classifier.js:2020]: Should this be `initOnOffButtons` now?

**Review by mrstegeman — CHANGES_REQUESTED** [review]

File: zb-classifier.js (hunk 2 above)
Comment (mrsteakhouse) [reply zb-classifier.js:495]: done

File: zb-classifier.js (hunk 3 above)
Comment (mrsteakhouse) [reply zb-classifier.js:526]: done

File: zb-classifier.js (hunk 4 above)
Comment (mrsteakhouse) [reply zb-classifier.js:561]: done

File: zb-classifier.js (hunk 5 above)
Comment (mrsteakhouse) [reply zb-classifier.js:1765]: done

File: zb-classifier.js (hunk 6 above)
Comment (mrsteakhouse) [reply zb-classifier.js:2135]: done

File: zb-classifier.js (hunk 7 above)
Comment (mrsteakhouse) [reply zb-classifier.js:2060]: renames

File: zb-classifier.js (hunk 8 above)
Comment (mrsteakhouse) [reply zb-classifier.js:2020]: renamed. thank you for taking the time.

File: zb-classifier.js (hunk 1 above)
Comment (mrstegeman) [reply zb-classifier.js:1808]: This still applies, inside the `if` block conditions.

**Comment by mrstegeman** [comment]
@mrsteakhouse just waiting on the one small spacing change.

File: zb-classifier.js (hunk 1 above)
Comment (mrsteakhouse) [reply zb-classifier.js]: forgot about this one. fixed now.

**Review by mrstegeman — APPROVED** [review]

**Comment by chas-iot** [comment]
The commit https://github.com/WebThingsIO/zigbee-adapter/commit/4518496ba207cd1649da8512e1c833693869e04e means that the adapter no longer crashes. However, the Tradfri remote control with deviceId = 0820 no longer has a scene property and the buttons do not work.
Initial debugging indicates that
```javascript
const genScenesOutputEndpoints =
            node.findZhaEndpointWithOutputClusterIdHex(
              CLUSTER_ID.GENSCENES_HEX);
```
results in null whereas previous to this PR it did return an end point.
I accept that the physical device appears to be buggy as it does not initially report the Cluster Id for Scenes however later it sends messages to that Cluster Id. Which means that a good fix may be difficult. 
I will look into this as my other commitments allow.

**Comment by chas-iot** [comment]
see [Zigbee adaptor crashing](https://discourse.mozilla.org/t/zigbee-adaptor-crashing/70175/7)
I'm requesting this PR be reverted until the collateral damage is resolved.

**Pull request merged**